runtimes/protocol/identity-management.ts

```diff
@@ -9,49 +10,99 @@ export type E_CANNOT_WRITE_SHARED_CONFIG = 'E_CANNOT_WRITE_SHARED_CONFIG'
+    E_CANNOT_WRITE_SSO_CACHE: 'E_CANNOT_WRITE_SSO_CACHE',
+    E_PROFILE_NOT_FOUND: 'E_PROFILE_NOT_FOUND',
+    E_CANNOT_CREATE_PROFILE: 'E_CANNOT_CREATE_PROFILE',
+    E_CANNOT_OVERWRITE_PROFILE: 'E_CANNOT_OVERWRITE_PROFILE',
+    E_INVALID_PROFILE: 'E_INVALID_PROFILE',
+    E_SSO_SESSION_NOT_FOUND: 'E_SSO_SESSION_NOT_FOUND',
+    E_CANNOT_CREATE_SSO_SESSION: 'E_CANNOT_CREATE_SSO_SESSION',
+    E_CANNOT_OVERWRITE_SSO_SESSION: 'E_CANNOT_OVERWRITE_SSO_SESSION',
+    E_INVALID_SSO_SESSION: 'E_INVALID_SSO_SESSION',
+    E_INVALID_TOKEN: 'E_INVALID_TOKEN',
+} as const
+
+export interface AwsResponseErrorData {
+    awsErrorCode: string
+}
+
+export class AwsResponseError<D extends AwsResponseErrorData> extends ResponseError<D> {
+    constructor(messageOrError: unknown, data: D, code: number = LSPErrorCodes.RequestFailed) {
+        super(code, (messageOrError as object)?.toString(), data)
+    }
+}
 
-export type ProfileKind = SsoTokenProfileKind
+// listProfiles
+export type ProfileKind = 'Unknown' | 'SsoTokenProfile'
 
 export const ProfileKind = {
-    SsoToken: 'SsoToken',
+    SsoTokenProfile: 'SsoTokenProfile',
+    Unknown: 'Unknown',
 } as const
 
 export interface Section {
     name: string
 }
 
-export interface Profile extends Section {
-    readonly kind: ProfileKind
-    region?: string
+// Profile and SsoSession use 'settings' property as namescope for their settings to avoid future
+// name conflicts with 'kind' and 'name' properties as well as making some setting iteration operations
+// easier.
+
+export interface Profile {
+    kind: ProfileKind
+    name: string
+    settings: {
+        region?: string
+        sso_session?: string
+    }
 }
 
-export interface SsoTokenProfile extends Profile {
-    readonly kind: SsoTokenProfileKind
-    ssoSessionName: string
+export interface SsoSession {
+    name: string
+    settings: {
+        sso_start_url: string
+        sso_region: string
+        sso_registration_scopes?: string[]
+    }
 }
 
-export interface SsoSession extends Section {
-    ssoStartUrl: string
-    ssoRegion: string
-    ssoRegistrationScopes?: string[]
+export type ListProfilesParams = {
+    // Intentionally left blank
 }
 
 export interface ListProfilesResult {
-    profiles?: (Profile | SsoTokenProfile)[]
-    ssoSessions?: SsoSession[]
+    profiles: Profile[]
+    ssoSessions: SsoSession[]
 }
-export interface ListProfilesError {
-    errorCode: E_UNKNOWN | E_TIMEOUT | E_RUNTIME_NOT_SUPPORTED | E_CANNOT_READ_SHARED_CONFIG
+
+export interface ListProfilesErrorData extends AwsResponseErrorData {
+    awsErrorCode: E_UNKNOWN | E_TIMEOUT | E_RUNTIME_NOT_SUPPORTED | E_CANNOT_READ_SHARED_CONFIG
+}
+
+export class ListProfilesError extends AwsResponseError<ListProfilesErrorData> {
+    constructor(messageOrError: unknown, data: ListProfilesErrorData, code: number = LSPErrorCodes.RequestFailed) {
+        super(messageOrError, data, code)
+        Object.setPrototypeOf(this, new.target.prototype)
```

Comment:
It is required for subclassing some built-in types, in this case Error, in TypeScript for the correct type to be assigned to this (e.g. to get back ListProfilesError rather than Error on instanceof)

https://github.com/microsoft/TypeScript/wiki/FAQ#why-doesnt-extending-built-ins-like-error-array-and-map-work